### PR TITLE
Passing dependencies to Application via constructor.

### PR DESCRIPTION
**Pull request by kiminuo**
## What does the pull request do?

Attempts to allow passing dependencies to applications without using `static` fields.

**This is a draft PR for the sake of starting discussion about this topic.**

## What is the current behavior?

`TApp` is always created using parameter-less constructor.

## What is the updated/expected behavior with this PR?

User can specify her own `TApp` factory function so that `TApp` can receive all dependencies through a constructor.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes

I think it should not break any existing code.

## Fixed issues

None.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4514) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4514) before we can accept your contribution.<br/><br/>:x: kiminuo [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4514)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=4514) it.</sub>

**Review by danwalmsley — CHANGES_REQUESTED** [review]
Please remove all the unrealted changed and formatting... so the diff only shows the functional changes for the new api.

**Comment by kiminuo** [comment]
@danwalmsley Better now?

File: tests/Avalonia.Controls.UnitTests/AppBuilderTests.cs
Comment (danwalmsley) [inline tests/Avalonia.Controls.UnitTests/AppBuilderTests.cs:24]: you can have `private AppWithDependencies ()` here rather than an exception... this makes it impossible for anyone to call the default ctor.

File: tests/Avalonia.Controls.UnitTests/AppBuilderTests.cs
Comment (danwalmsley) [inline tests/Avalonia.Controls.UnitTests/AppBuilderTests.cs:34]: nit: missing newline after }

**Review by danwalmsley — CHANGES_REQUESTED** [review]
couple of minor nits.

Looks good, will get @grokys or @kekekeks to give some input also.

**Review by danwalmsley — APPROVED** [review]
LGTM

**Comment by workgroupengineering** [comment]
Please consider adding documentation at https://github.com/AvaloniaUI/Avaloniaui.net.

**Review by grokys — APPROVED** [review]
I think this makes sense, though I'd wait for @kekekeks' approval as he designed the app lifetimes stuff.

**Comment by danwalmsley** [comment]
@kekekeks gave this the thumbs up!

**Pull request merged**
